libs/designer/src/lib/core/state/tokens/tokenSelectors.ts

```diff
@@ -0,0 +1,22 @@
+import type { RootState } from '../../store';
+import type { TokensState } from './tokensSlice';
+import { createSelector } from '@reduxjs/toolkit';
+import { useSelector } from 'react-redux';
+
+export const getTokenState = (state: RootState): TokensState => state.tokens;
+
+export const useUpstreamNodes = (id?: string) => {
+  return useSelector(
+    createSelector(getTokenState, (state: TokensState) => {
+      const upstreamNodes = new Set();
+      const nodeTokens = state.outputTokens[id ?? ''];
+      // TODO: Support variables
+      const upstreamNodeIds = nodeTokens?.upstreamNodeIds ?? [];
+
+      upstreamNodeIds.forEach((upstreamNodeId) => {
+        upstreamNodes.add(upstreamNodeId);
+      });
```

Comment:
If you're just trying to remove duplicates, instead of making the set obj and building it with `.add` you could just do this
```suggestion
      const nodeTokens = state.outputTokens[id ?? ''];
      // TODO: Support variables
      return [...(new Set(nodeTokens?.upstreamNodeIds ?? []))];
```
We could honestly just make it a `RemoveDuplicates` utils function for arrays